sonic_data_client/mixed_db_client.go

```diff
@@ -75,6 +75,87 @@ type MixedDbClient struct {
 
 var mixedDbClientMap = map[string]MixedDbClient{}
 
+func hget(table swsscommon.Table, key string, field string) (string, error) {
+	var fieldValuePairs = swsscommon.NewFieldValuePairs()
+	var result = table.Get(key, fieldValuePairs)
+	if result {
+		var fieldCount = int(fieldValuePairs.Size())
+		for idx := 0; idx < fieldCount; idx++ {
+			var pair = fieldValuePairs.Get(idx)
+			if pair.GetFirst() == field {
+				return pair.GetSecond(), nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Can't read %s:%s from %s table", key, field, table.GetTableName())
+}
+
+func getDpuAddress(dpuId string) (string, error) {
+	var configDb = swsscommon.NewDBConnector("CONFIG_DB", SWSS_TIMEOUT, false)
+
+	// Find DPU address by DPU ID from CONFIG_DB
+	// Design doc: https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/ip-address-assigment/smart-switch-ip-address-assignment.md?plain=1
+
+	// get bridge plane
+	var bridgeTable = swsscommon.NewTable(configDb, "MID_PLANE_BRIDGE")
+	bridgePlane, err := hget(bridgeTable, "GLOBAL", "bridge");
+	if err != nil {
+		return "", err
+	}
+
+	// get DPU interface by DPU ID
+	var dpuTable = swsscommon.NewTable(configDb, "DPUS")
+	dpuInterface, err := hget(dpuTable, dpuId, "midplane_interface");
+	if err != nil {
+		return "", err
+	}
+
+	// get DPR address by DPU ID and brdige plane
+	var dhcpPortTable = swsscommon.NewTable(configDb, "DHCP_SERVER_IPV4_PORT")
+	var dhcpPortKey = bridgePlane + "|" + dpuInterface
+	dpuAddresses, err := hget(dhcpPortTable, dhcpPortKey, "ips");
+	if err != nil {
+		return "", err
+	}
+
+	var dpuAddressArray = strings.Split(dpuAddresses, ",")
+	if len(dpuAddressArray) == 0 {
+		return "", fmt.Errorf("Can't find address of dpu:'%s' from DHCP_SERVER_IPV4_PORT table", dpuId)
+	}
+
+	return dpuAddressArray[0], nil
+}
+
+func getZmqAddress(container string, zmqPort string) (string) {
+	var dpuAddress, err = getDpuAddress(container)
+	if err != nil {
+		// When can't find address by container, 'container' is not a DPU ID
+		dpuAddress = "127.0.0.1"
```

Comment:
>dpuAddress [](http://example.com/codeflow?start=2&length=10)

If there is err, should we keep return err to outside?

Reply:
According to current GNMI API design, the 'container' can be namespace or DPU, the only way to identify it is check if the 'container' has an address assigned in DHCP, if 'container' is a namespace, we need return local address.

Reply:
If getDpuAddress return any err, getZmqAddress should return err too.

Reply:
Will throw error when get DPU failed

Reply:
Still applicable.

Reply:
Fixed.